### DISCUSSION TITLE
Can you set the colour of individual LEDs based on their physical location along the string?

### DISCUSSION THREAD

**Discussion by MaxPalmer-UH** [Anders-Holst/xled_plus#18] [Q&A] [answered]
Is there a way of setting the colour of individual LEDs based on their physical location along the device. Let's say I have a cluster set, with LEDs in a linear array along the length of the string. Does the hardware address each LED based on an index into it's physical position along the string, does it know this information or does it require calibrating the device when in a line/unrolled first.

Could I set every fourth LED in the string to a particular colour? etc?

Can this be done using this API?

Thanks.

Max

**Comment by Anders-Holst** [comment, marked as answer]
Yes it is indeed possible.

The function `make_func_pattern` in HighControlInterface produces a pattern based on the led index along the string. It takes a function as argument, which in turn takes the index of one led and returns the color of that specific led as an (r,g,b)-tuple. 

Note that many devices have two strings emanating from the device, and the ordering of the leds start at one string closest to the device and goes to the tip of that string, then starts from closest the device on the other string and goes to the tip. So if you arrange them in a line with the device in the middle, or in a circle where the two tips touch each other, the indices would make a jump in the middle, from the tip of on string to the base of the other. Then you can set the optional argument `circular` to True when calling `make_func_pattern`, and it will rearrange the led indices so they instead start at one tip, goes in towards the device, and then travels from the device out to the tip on the other string.

The specific pattern you are mentioning, to set every fourth led to the same color, is actually already provided in "pattern.py": The function `make_alternating_color_pattern` takes the HighControlInterface object of you device, and a list (of some length) with RGB-tuples, and alternates those colors along the string. There are several other examples in that file which you can check out.

**Reply by DrMaxP** [reply]
Thanks for such a detailed reply. Amazing
